### PR TITLE
Use immer instead of rust

### PR DESCRIPTION
**Pull request by dwightguth**
Removes rust from the codebase and replaces references everywhere to the C++ immer library. I tested the performance against KEVM and saw a roughly 25% speedup in the codebase resulting from this. All tests pass as well.

@chathhorn can you please pull the latest changes of this branch and test that there are no regressions against the rv-match codebase?

### CODE REVIEW

**Comment by chathhorn** [comment]
I tested f0f852f with the C semantics unit tests and don't see any regressions.

File: runtime/meta/substitution.cpp [hunk 1]
```diff
@@ -3,6 +3,7 @@
 #include "runtime/header.h"
 #include "runtime/alloc.h"
 
+
```
Comment (mickyabir) [inline runtime/meta/substitution.cpp]: am I missing something here?

File: unittests/runtime-collections/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,12 @@
+add_kllvm_unittest(runtime-collections-tests
+  lists.cpp
+  maps.cpp
+  main.cpp
```
Comment (mickyabir) [inline unittests/runtime-collections/CMakeLists.txt]: what about the `sets` tests?

**Comment by mickyabir** [comment]
Approved but please look at the changes I made

**Pull request merged**